src/Microsoft.ML.Dnn/LearningRateScheduler.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Transforms
+{
+    /// <summary>
+    /// A class that contains the current train state to use for learning rate scheduling.
+    /// </summary>
+    internal class TrainState
+    {
+        internal int CurrentBatchIndex;
+        internal int CurrentEpoch;
+        internal int BatchSize;
+        internal int BatchesPerEpoch;
+    }
+
+    /// <summary>
+    /// This interface defines a learning rate scheduler.
+    /// </summary>
+    public abstract class LearningRateScheduler
+    {
+        internal abstract float GetLearningRate(TrainState options);
+    };
+
+    /// <summary>
+    /// This class implements linear scaling rule and LR decay.
+    /// Implementation adopted from RESNET-CIFAR benchmark test in Tensorflow slim.
+    /// https://github.com/tensorflow/models/blob/b974c3f95a37acedcc3c58566834c78fcae4b214/official/vision/image_classification/resnet_cifar_main.py
+    /// </summary>
+    public sealed class LsrDecay : LearningRateScheduler
+    {
+        /// <summary>
+        /// Learning rate is scaled at epoch boundaries provided in LrSchedule to corresponding multiplier in the LrSchedule.
+        /// Format for LrSchedule: {epoch, scaling factor}
+        /// </summary>
+        public readonly float[,] LrSchedule;
```

Comment:
A thing with having public array fields is that people can change the elements of the array. Even though it is `readonly`, that only means that someone can't change this reference to a different array. But the array itself can be changed.

Reply:
Also having a multi-dimensional array `[,]` in a public API is a little bit confusing. If you used an `IReadOnlyList` of some new `struct` type, then you can name each argument:

```
public readonly struct LearningDecayItem
{
    public readonly float Epoch;  // actually should this be float or int?
    public readonly float ScalingFactor;
}
```

Then people know which is which easily.

Reply:
What remedy do you suggest? Basically I want it to be set only in the constructor.

Reply:
Thanks will change that.